ml/nml.cc

```diff
@@ -0,0 +1,1133 @@
+    const char *mt_str = nml_metric_type_to_string(dim->mt);
+    const char *ts_str = nml_training_status_to_string(dim->ts);
+    const char *tr_str = nml_training_result_to_string(dim->tr.result);
+
+    const char *fmt =
+        "[ML] %s.%s: MLS=%s, MT=%s, TS=%s, Result=%s, "
+        "ReqTime=%ld, FEOReq=%ld, LEOReq=%ld, "
+        "FEOResp=%ld, LEOResp=%ld, QTR=<%ld, %ld>, DBTR=<%ld, %ld>, Collected=%zu, Total=%zu";
+
+    error(fmt,
+          chart_id, dimension_id, mls_str, mt_str, ts_str, tr_str,
+          dim->tr.request_time, dim->tr.first_entry_on_request, dim->tr.last_entry_on_request,
+          dim->tr.first_entry_on_response, dim->tr.last_entry_on_response,
+          dim->tr.query_after_t, dim->tr.query_before_t, dim->tr.db_after_t, dim->tr.db_before_t, dim->tr.collected_values, dim->tr.total_values
+    );
+}
+
+nml_dimension_t *nml_dimension_new(RRDDIM *rd) {
+    nml_dimension_t *dim = new nml_dimension_t();
+
+    dim->rd = rd;
+
+    dim->mt = METRIC_TYPE_CONSTANT;
+    dim->ts = TRAINING_STATUS_UNTRAINED;
+
+    dim->last_training_time = 0;
+
+    nml_kmeans_init(&dim->kmeans, 2, 1000);
+
+    if (simple_pattern_matches(Cfg.sp_charts_to_skip, rrdset_name(rd->rrdset)))
+        dim->mls = MACHINE_LEARNING_STATUS_DISABLED_DUE_TO_EXCLUDED_CHART;
+    else
+        dim->mls = MACHINE_LEARNING_STATUS_ENABLED;
+
+    netdata_mutex_init(&dim->mutex);
+
+    dim->km_contexts.reserve(Cfg.num_models_to_use);
+
+    return dim;
+}
+
+void nml_dimension_delete(nml_dimension_t *dim) {
+    netdata_mutex_destroy(&dim->mutex);
+    delete dim;
+}
+
+nml_chart_t *nml_chart_new(RRDSET *rs) {
+    nml_chart_t *chart = new nml_chart_t();
+
+    chart->rs = rs;
+    chart->mls = nml_machine_learning_stats_t();
+
+    netdata_mutex_init(&chart->mutex);
+
+    return chart;
+}
+
+void nml_chart_delete(nml_chart_t *chart) {
+    netdata_mutex_destroy(&chart->mutex);
+    delete chart;
+}
+
+static bool nml_chart_is_available_for_ml(nml_chart_t *chart) {
+    return rrdset_is_available_for_exporting_and_alarms(chart->rs);
+}
+
+static std::string ml_dimension_get_id(RRDDIM *rd) {
+    RRDSET *rs = rd->rrdset;
+
+    std::stringstream ss;
+    ss << rrdset_context(rs) << "|" << rrdset_id(rs) << "|" << rrddim_name(rd);
+    return ss.str();
+}
+
+static void nml_chart_get_models_as_json(nml_chart_t *chart, nlohmann::json &j) {
+    netdata_mutex_lock(&chart->mutex);
+
+    void *rdp = NULL;
+    rrddim_foreach_read(rdp, chart->rs) {
+        RRDDIM *rd = static_cast<RRDDIM *>(rdp);
+        nml_dimension_t *dim = reinterpret_cast<nml_dimension_t *>(rd->ml_dimension);
+        if (!dim)
+            continue;
+
+        nlohmann::json jarray = nlohmann::json::array();
+#if 0
+        for (const KMeans &KM : nml_dimension_models(D)) {
+            nlohmann::json tmp;
+
+            KM.toJson(tmp);
+            jarray.push_back(tmp);
+            j[ml_dimension_get_id(D->rd)] = jarray;
+        }
+#else
```

Comment:
Is this required or can it be deleted?

Reply:
It can be deleted, but I'd like to keep it for a couple days as it is because it's will be used for testing between I and Andrew.

Reply:
OK fair enough.